sklearn/cluster/optics_.py

```diff
@@ -527,3 +553,362 @@ def cluster_optics_dbscan(reachability, core_distances, ordering, eps=0.5):
+        The list of clusters in the form of [start, end] in each row, with all
+        indices inclusive. The clusters are ordered in a way that larger
+        clusters encompassing smaller clusters come after those smaller
+        clusters.
+    """
+
+    # all indices are inclusive (specially at the end)
+    n_samples = len(reachability_plot)
+    # add an inf to the end of reachability plot
+    # but process the data only until the last actual point
+    # this is fine since the last point is considered upward anyway
+    reachability_plot = np.hstack((reachability_plot, np.inf))
+
+    if min_cluster_size <= 1:
+        min_cluster_size = max(2, min_cluster_size * n_samples)
+
+    xi_complement = 1 - xi
+    sdas = []  # steep down areas, introduced in section 4.3.2 of the paper
+    clusters = []
+    index = 0
+    mib = 0.  # maximum in between
+
+    ratio = reachability_plot[:-1] / reachability_plot[1:]
+    steep_upward = ratio <= xi_complement
+    steep_downward = ratio >= 1 / xi_complement
+    steep = steep_downward.astype(int) - steep_upward.astype(int)
+    downward = ratio > 1
+    upward = ratio < 1
+
+    # the following loop is is almost exactly as Figure 19 of the paper.
+    # it jumps over the areas which are not either steep down or up areas
+    for steep_index in iter(np.flatnonzero(steep)):
+        # just continue if steep_index has been a part of a discovered xward
+        # area.
+        if steep_index < index:
+            continue
+
+        mib = max(mib, np.max(reachability_plot[index:steep_index + 1]))
+        index = steep_index
+
+        # check if a steep downward area starts
+        if steep_downward[index]:
+            sdas = _update_filter_sdas(sdas, mib, xi_complement)
+            D_start = index
+            index, end = _extend_region(steep_downward, upward, D_start,
+                                        min_samples, n_samples)
+            D = _Area(start=D_start, end=end,
+                      maximum=reachability_plot[D_start], mib=0.)
+            sdas.append(D)
+            mib = reachability_plot[end + 1]
+
+        elif steep_upward[index]:
+            sdas = _update_filter_sdas(sdas, mib, xi_complement)
+            U_start = index
+            index, end = _extend_region(steep_upward, downward, U_start,
+                                        min_samples, n_samples)
+            U = _Area(start=U_start, end=end, maximum=reachability_plot[end],
+                      mib=-1)
+            mib = reachability_plot[end + 1]
+
+            U_clusters = []
+            for D in sdas:
+                c_start = D.start
+                c_end = min(U.end, n_samples - 1)
+
+                # line (**)
+                if reachability_plot[c_end + 1] * xi_complement < D.mib:
+                    continue
+
+                # 4
+                if D.maximum * xi_complement >= reachability_plot[c_end + 1]:
+                    while (reachability_plot[c_start + 1] >
```

Comment:
this could perhaps do with a comment (or you could use a searchsorted between c_start and the argmin of the D)

Reply:
reachability_plot is not a sorted array though. Added a comment.